### ISSUE TITLE
Emulate network and CPU throttling

### ISSUE THREAD

**Issue by dlockhart** [puppeteer/puppeteer#764]
Feature request probably related to #63, which is blocked on [the ability to emulate network conditions](https://bugs.chromium.org/p/chromium/issues/detail?id=728451) in headless Chrome. Feel free to close if you feel that issue is a better fit, but I wanted to specifically call out CPU throttling and non-offline network throttling, which don't seem to be captured there.

When running performance tests and capturing various marks (FP, FCP, FMP one-day-soon-maybe, TTI, etc.), it would be swell if I could emulate different network (offline, slow 3G, fast 3G, etc.) and CPU (2x slower, 5x slower, etc.) throttling configurations -- in the same way Chrome dev tools does.

That way I could combine different headless runs with different configurations into a distribution and create a more "real world" picture of how my application performs in the wild.

**Comment by JoelEinbinder** [comment]
These all seem important, but I am struggling to come up with a way that Puppeteer would provide value over [Lighthouse](https://github.com/GoogleChrome/lighthouse) in this area. Can you elaborate on your use case?

**Comment by Garbee** [comment]
This was already covered in lesser detail in the comments to #63.

Emulation is nothing like "real world" scenarios. There are too many factors with devices (especially mobile) that affects their performance. The only way to get accurate numbers for non-desktop devices is to physically run on those devices. Just a few of those factors are:

- CPU
  - Core count
  - Architecture
  - Load existing from OS
  - Thermals at run-time affecting performance
- RAM
  - Size
  - Access speed
  - Load existing from OS
- Network
  - Latency to towers/radios
  - Current traffic patterns throughout the network of that carrier (if not WiFi)
  - Current radio activity (If the radio is doing a cold start things take a few hundred MS longer in most *best* cases)

These factors and more vary so widely between devices and depend upon the current state of the device itself. Emulation within DevTools is simply a good **guideline** that you can use while developing without constantly hitting back to a normal device. When you're deciding on whether or not your build is good enough to deploy, physical device testing is the way to go.

IMO having deeper emulation conditions within Puppeteer isn't as useful as developers may think it is. Instead of including it an allowing the notion to further perpetuate that emulation provides accurate results, let's not include it. Emulation is extremely useful for active development in a handful of scenarios. But, for "real world" performance metrics it can never be an ideal testing scenario to make decisions on.

**Comment by dlockhart** [comment]
Thank you both your your responses. @Garbee you've convinced me that emulation isn't going to provide the accuracy or consistency I'm looking for in a CI environment to compare test runs over time. Will investigate using Lighthouse's CLI or an out-of-the-box solution like [Calibre](https://building.calibreapp.com/). Closing.

**Comment by Garbee** [comment]
Lighthouse is using the same Chrome internal network throttling. So for "real world" cases it once again doesn't cut the mustard as it were. It is still a good guideline for development, but it isn't what you're looking for to test what actual network conditions could be like. Since latency is statically emulated, there is no jitter added, radio timing isn't taken into account, etc.
